homeassistant/components/sonarr/sensor.py

```diff
@@ -72,198 +61,430 @@
+            usage = free / total * 100
+
+            attrs[
+                disk.path
+            ] = f"{free:.2f}/{total:.2f}{self._unit_of_measurement} ({usage:.2f}%)"
+
+        return attrs
+
+    @property
+    def state(self) -> Union[None, str, int, float]:
+        """Return the state of the sensor."""
+        free = self._to_unit(self._total_free)
+        return f"{free:.2f}"
+
+
+class SonarrQueueSensor(SonarrSensor):
+    """Defines a Sonarr Queue sensor."""
 
-    def update(self):
-        """Update the data for the sensor."""
+    def __init__(self, sonarr: Sonarr, entry_id: str) -> None:
+        """Initialize Sonarr Queue sensor."""
+        self._queue = []
+
+        super().__init__(
+            sonarr=sonarr,
+            entry_id=entry_id,
+            icon="mdi:download",
+            key="queue",
+            name=f"{sonarr.app.info.app_name} Queue",
+            unit_of_measurement="Episodes",
+            enabled_default=False,
+        )
+
+    @sonarr_exception_handler
+    async def async_update(self) -> None:
+        """Update entity."""
+        self._queue = await self.sonarr.queue()
+
+    @property
+    def device_state_attributes(self) -> Optional[Dict[str, Any]]:
+        """Return the state attributes of the entity."""
+        attrs = {}
+
+        for item in self._queue:
+            remaining = 1 if item.size == 0 else item.size_remaining / item.size
+            remaining_pct = 100 * (1 - remaining)
+            name = f"{item.episode.series.title} {item.episode.identifier}"
+            attrs[name] = f"{remaining_pct:.2f}%"
+
+        return attrs
+
+    @property
+    def state(self) -> Union[None, str, int, float]:
+        """Return the state of the sensor."""
+        return len(self._queue)
+
+
+class SonarrSeriesSensor(SonarrSensor):
+    """Defines a Sonarr Series sensor."""
+
+    def __init__(self, sonarr: Sonarr, entry_id: str) -> None:
+        """Initialize Sonarr Series sensor."""
+        self._items = []
+
+        super().__init__(
+            sonarr=sonarr,
+            entry_id=entry_id,
+            icon="mdi:television",
+            key="series",
+            name=f"{sonarr.app.info.app_name} Shows",
+            unit_of_measurement="Series",
+            enabled_default=False,
+        )
+
+    @sonarr_exception_handler
+    async def async_update(self) -> None:
+        """Update entity."""
+        self._items = await self.sonarr.series()
+
+    @property
+    def device_state_attributes(self) -> Optional[Dict[str, Any]]:
+        """Return the state attributes of the entity."""
+        attrs = {}
+
+        for item in self._items:
+            attrs[item.series.title] = f"{item.downloaded}/{item.episodes} Episodes"
+
+        return attrs
+
+    @property
+    def state(self) -> Union[None, str, int, float]:
+        """Return the state of the sensor."""
+        return len(self._items)
+
+
+class SonarrUpcomingSensor(SonarrSensor):
```

Comment:
I know that this sensor is used for a Lovelace card, but I think that it would be better if we make this a websocket API ? Really no reason for it to be an entity except for our state machine being abused as a data storage by the LL card ;)

Reply:
it did seem there should be something better but is it possible to move forward with such a change in iterations? Im not familiar with WS API so it would take some research and experimentation when ive already done a lot of work on modernizing here

Reply:
I dont think this sensor is used for https://github.com/custom-cards/upcoming-media-card if thats what you mean. Seems theres a custom component for that: https://github.com/custom-components/sensor.sonarr_upcoming_media